### PR TITLE
Updates the ensembling HOWTO to Optax.

### PR DESCRIPTION
**Pull request by andsteing**
This PR also bases the code on the more recent version of the MNIST example and
implements the actual ensembling (before this commit, the individual
members' predictions weren't actually ensembled, but instead the accuracy
was incorrectly averaged from individual members' accuracies).

TESTED_MANUALLY:
https://colab.research.google.com/drive/1Hgc2yYdMbrVNEELqZtF7vr6aIYF3WknS

### CODE REVIEW

**Comment by andsteing** [comment]
#1357

File: docs/howtos/ensembling.rst [hunk 1]
```diff
@@ -1,307 +1,274 @@
...
 
-  def train_epoch(optimizer, train_ds, rng, batch_size=10):
+  def train_epoch(state, train_ds, batch_size, rng):
     train_ds_size = len(train_ds['image'])
     steps_per_epoch = train_ds_size // batch_size
 
-    perms = random.permutation(rng, len(train_ds['image']))
-    perms = perms[:steps_per_epoch * batch_size]
+    perms = jax.random.permutation(rng, len(train_ds['image']))
+    perms = perms[:steps_per_epoch * batch_size]
```
Comment (marcvanzee) [inline docs/howtos/ensembling.rst:165]:
```suggestion
    perms = perms[:steps_per_epoch * batch_size]
```

File: docs/howtos/ensembling.rst [hunk 2]
```diff
@@ -1,307 +1,274 @@
...
-  def train_step(optimizer, batch):
-    """Train for a single step."""
+  @jax.jit  #!
+  def apply_model(state, images, labels):
     def loss_fn(params):
-      logits = CNN().apply({'params': params}, batch['image'])
-      loss = cross_entropy_loss(logits, batch['label'])
+      logits = CNN().apply({'params': params}, images)
+      one_hot = jax.nn.one_hot(labels, 10)
+      loss = jnp.mean(optax.softmax_cross_entropy(logits=logits, labels=one_hot))
```
Comment (marcvanzee) [inline docs/howtos/ensembling.rst:118]: Nit: You could consider breaking this up in two lines since otherwise the side-by-side is a bit too wide.

**Review by marcvanzee — APPROVED** [review]
Looks great, thanks a lot for improving this HOWTO and catching that we actually were doing ensembling incorrectly!

One note: you now use the full MNIST model, I wonder whether this will slow down the tests a lot, rather than using a tiny model (as I did initially). You could just try running a quick comparison on your Colab on CPU and see whether it matters. If not then no problem, but if it leads to a significant slowdown then this may become annoying when we want to run our tests locally.

File: docs/howtos/ensembling.rst [hunk 3]
```diff
@@ -1,307 +1,274 @@
 Ensembling on multiple devices
-=============================
+==============================
```
[the remark this answers is not in the document]
Comment (andsteing) [reply docs/howtos/ensembling.rst:2]: Yes!

File: docs/howtos/ensembling.rst [hunk 4]
```diff
@@ -1,307 +1,274 @@
...
 
 We show how to train an ensemble of CNNs on the MNIST dataset, where the size of
 the ensemble is equal to the number of available devices. In short, this change
 be described as: 
 
-* make a number of functions parallel using ``jax.pmap``, 
-* replicate the inputs carefully,
-* make sure the parallel and non-parallel logic interacts correctly.
+* make a number of functions parallel using |jax.pmap()|_,
+* split the random seed to have different parameter initialization,
```
[the remark this answers is not in the document]
Comment (andsteing) [reply docs/howtos/ensembling.rst:9]: Done.

File: docs/howtos/ensembling.rst (hunk 1 above)
Comment (andsteing) [reply docs/howtos/ensembling.rst:165]: Done

File: docs/howtos/ensembling.rst [hunk 5]
```diff
@@ -1,307 +1,274 @@
...
+    return state.apply_gradients(grads=grads)
 
 Training the Ensemble
---------------------------------
+---------------------
 
-Next we transform the ``train_epoch`` function.
+Next we transform the ``train_epoch()`` function. When calling the pmapped
+functions from above, we mainly need to take care of duplicating the arguments
+for all devices where necessary, and de-duplicating the
```
[the remark this answers is not in the document]
Comment (andsteing) [reply docs/howtos/ensembling.rst:154]: Done.

File: docs/howtos/ensembling.rst (hunk 2 above)
Comment (andsteing) [reply docs/howtos/ensembling.rst:118]: Done.

**Comment by andsteing** [comment]
Thanks for the quick review!

wrt performance: I did timing measurements on my computer and it's exactly the same (note the small batch size and the single epoch) ... but thanks for thinking of this!

**Review by marcvanzee — APPROVED** [review]
